// Copyright (c) 2015-2019 Nikos Nikoleris
// All rights reserved
//
// The license below extends only to copyright in the software and shall
// not be construed as granting a license to any other intellectual
// property including but not limited to intellectual property relating
// to a hardware implementation of the functionality of the software
// licensed hereunder.  You may use the software subject to the license
// terms below provided that you ensure that this notice is replicated
// unmodified and in its entirety in all distributions of the software,
// modified or unmodified, in source code or in binary form.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met: redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer;
// redistributions in binary form must reproduce the above copyright
// notice, this list of conditions and the following disclaimer in the
// documentation and/or other materials provided with the distribution;
// neither the name of the copyright holders nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
// Authors: Nikos Nikoleris

syntax = "proto2";

// Put all the generated messages in a namespace
package ProtoReuseProfile;

message Reuse {
  required uint64 dist = 1;
  required float freq = 2;
}

message InstReuseHist {
  required uint64 addr = 1;
  repeated Reuse reuse = 2;
}

message ReuseHist {
  repeated InstReuseHist instReuseHist = 1;
}

message Window {
  required uint64 id = 1;
  repeated Reuse reuse = 2;
}

message Vicinity {
  repeated Window window = 1;
  required uint64 windowSize = 2;
}

message SampledInst {
  required uint64 addr = 1;
  required uint64 freq = 2;
}

message SampledInsts {
  repeated SampledInst sampledInst = 1;
}

message Access {
  optional uint64 pc = 1;
  optional uint64 addr = 2;
  optional uint64 physAddr = 3;
  optional uint64 size = 4;
  optional uint64 dist = 5;
}

message Accesses {
  repeated Access accesses = 1;
}

message ReuseProfile {
  required Vicinity vicinity = 1;
  optional ReuseHist reuseHist = 2;
  optional SampledInsts sampledInsts = 3;
  repeated Access accesses = 4;
}

